File: python/solvers/fv/fluxes.py
from numpy import complex128, dot, zeros
from scipy.linalg import eig, solve

from solvers.basis import NODES, WGHTS


def B_INT(obj, qL, qR, d):
    """ Returns the jump matrix for B, in the dth direction.
    """
    ret = zeros(obj.NV)
    Δq = qR - qL
    for i in range(obj.N):
        q = qL + NODES[i] * Δq
        B = obj.nonconservative_matrix(q, d, obj.model_params)
        ret += WGHTS[i] * dot(B, Δq)
    return ret


def D_OSH(obj, qL, qR, d):
    """ Returns the Osher flux component, in the dth direction
    """
    ret = zeros(obj.NV, dtype=complex128)
    Δq = qR - qL

    for i in range(obj.N):
        q = qL + NODES[i] * Δq
        M = obj.system_matrix(q, d, obj.model_params)
        λ, R = eig(M, overwrite_a=1)
        b = solve(R, Δq)
        ret += WGHTS[i] * dot(R, abs(λ) * b)

    return ret.real


def D_ROE(obj, qL, qR, d):
    """ Returns the Roe flux component, in the dth direction
    """
    M = zeros([obj.NV, obj.NV])
    Δq = qR - qL

    for i in range(obj.N):
        q = qL + NODES[i] * Δq
        M += WGHTS[i] * obj.system_matrix(q, d, obj.model_params)

    λ, R = eig(M, overwrite_a=1)
    b = solve(R, Δq)
    return dot(R, abs(λ) * b).real


def D_RUS(obj, qL, qR, d):
    """ Returns the Rusanov flux component, in the dth direction
    """
    max1 = obj.max_eig(qL, d, obj.model_params)
    max2 = obj.max_eig(qR, d, obj.model_params)
    return max(max1, max2) * (qR - qL)

File: cpp/src/solvers/fv/fluxes.cpp
#include "../../etc/globals.h"
#include "../../include/eigen3/Eigenvalues"
#include "../../system/eig.h"
#include "../../system/equations.h"
#include <cmath>

VecV Aint(VecVr qL, VecVr qR, int d, Par &MP) {
  // Returns the Osher-Solomon jump matrix for A, in the dth direction
  VeccV ret = VeccV::Zero();
  VecV Δq = qL - qR;
  VeccV Δqc = VeccV(Δq);
  Eigen::EigenSolver<MatV_V> es;

  VecV q;
  VeccV b;
  MatV_V J;
  for (int i = 0; i < N; i++) {
    q = qR + NODES(i) * Δq;
    J = system_matrix(q, d, MP);
    es.compute(J);
    b = es.eigenvectors().colPivHouseholderQr().solve(Δqc).array() *
        es.eigenvalues().array().abs();
    ret += WGHTS(i) * (es.eigenvectors() * b);
  }
  return ret.real();
}

VecV Bint(VecVr qL, VecVr qR, int d, Par &MP) {
  // Returns the jump matrix for B, in the dth direction.
  VecV ret = VecV::Zero();
  VecV Δq = qR - qL;
  VecV q, tmp;
  for (int i = 0; i < N; i++) {
    q = qL + NODES(i) * Δq;
    Bdot(tmp, q, Δq, d, MP);
    ret += WGHTS(i) * tmp;
  }
  return ret;
}

VecV Smax(VecVr qL, VecVr qR, int d, bool PERR_FROB, Par &MP) {
  double max1 = max_abs_eigs(qL, d, PERR_FROB, MP);
  double max2 = max_abs_eigs(qR, d, PERR_FROB, MP);
  return std::max(max1, max2) * (qL - qR);
}
